### PR TITLE
feat: use patches in scheduler to avoid races + add some more unit tests

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR updates the scheduler logic to use patches (instead of updates) to avoid races.

It also adds a number of unit tests that wasn't included in previous scheduler PRs.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

- [x] Unit tests

### CODE REVIEW

File: pkg/scheduler/framework/frameworkutils.go [hunk 1]
```diff
@@ -136,10 +150,13 @@ func crossReferencePickedCustersAndObsoleteBindings(
...
 		// Update the binding so that it is associated with the lastest scheduling policy.
-		binding.Spec.SchedulingPolicySnapshotName = policy.Name
+		updated.Spec.SchedulingPolicySnapshotName = policy.Name
 
 		// Add the binding to the toUpdate list.
-		toUpdate = append(toUpdate, binding)
+		toPatch = append(toPatch, &bindingWithPatch{
+			updated: updated,
+			patch:   patch,
+		})
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/frameworkutils.go:159]:
```suggestion
		toPatch = append(toPatch, &bindingWithPatch{
			updated: updated,
			patch:   client.MergeFrom(binding),
		})
```
nit, no need to use the variable or move the variable closer to where it's used

File: pkg/scheduler/framework/framework_test.go [hunk 2]
```diff
@@ -1393,3 +1448,205 @@ func TestCrossReferencePickedCustersAndObsoleteBindings(t *testing.T) {
...
+	}
+
+	// Check if the requested binding has been deleted.
+	deletedBinding := &fleetv1beta1.ClusterResourceBinding{}
+	if err := fakeClient.Get(ctx, types.NamespacedName{Name: anotherBindingName}, deletedBinding); err != nil {
+		t.Errorf("Get() binding %s = %v, want no error", anotherBindingName, err)
+	}
+	if deletedBinding.Spec.State != fleetv1beta1.BindingStateUnscheduled {
+		t.Errorf("binding %s state = %s, want unscheduled", anotherBindingName, deletedBinding.Spec.State)
+	}
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/framework_test.go:1651]: compare the full structure to test we only update the state and other fields should not change?

Comment (michaelawyu) [reply pkg/scheduler/framework/framework_test.go:1660]: Done.

**Review by zhiying-lin — APPROVED** [review]
LGTM :)

**Pull request merged**